service/matching/tasklist/adaptive_scaler.go

```diff
@@ -0,0 +1,255 @@
+		a.scope.IncCounter(metrics.CadenceFailures)
+	}
+}
+
+func (a *adaptiveScalerImpl) getPartitionConfig() *types.TaskListPartitionConfig {
+	partitionConfig := a.tlMgr.TaskListPartitionConfig()
+	if partitionConfig == nil {
+		partitionConfig = &types.TaskListPartitionConfig{
+			NumReadPartitions:  1,
+			NumWritePartitions: 1,
+		}
+	}
+	return partitionConfig
+}
+
+func (a *adaptiveScalerImpl) adjustWritePartitions(qps float64, numWritePartitions int32) int32 {
+	upscaleThreshold := float64(a.config.PartitionUpscaleRPS())
+	downscaleFactor := a.config.PartitionDownscaleFactor()
+	downscaleThreshold := float64(numWritePartitions-1) * upscaleThreshold * downscaleFactor / float64(numWritePartitions)
+	a.scope.UpdateGauge(metrics.EstimatedAddTaskQPSGauge, qps)
+	a.scope.UpdateGauge(metrics.TaskListPartitionUpscaleThresholdGauge, upscaleThreshold)
+	a.scope.UpdateGauge(metrics.TaskListPartitionDownscaleThresholdGauge, downscaleThreshold)
+
+	result := numWritePartitions
+	if qps > upscaleThreshold {
+		if !a.overLoad {
+			a.overLoad = true
+			a.overLoadStartTime = a.timeSource.Now()
+		} else if a.timeSource.Now().Sub(a.overLoadStartTime) > a.config.PartitionUpscaleSustainedDuration() {
+			result = getNumberOfPartitions(numWritePartitions, qps, upscaleThreshold)
+			a.overLoad = false
+		}
+	} else {
+		a.overLoad = false
+	}
+	if qps < downscaleThreshold {
+		if !a.underLoad {
+			a.underLoad = true
+			a.underLoadStartTime = a.timeSource.Now()
+		} else if a.timeSource.Now().Sub(a.underLoadStartTime) > a.config.PartitionDownscaleSustainedDuration() {
+			result = getNumberOfPartitions(numWritePartitions, qps, upscaleThreshold)
+			a.underLoad = false
+		}
+	} else {
+		a.underLoad = false
+	}
+	return result
+}
+
+func (a *adaptiveScalerImpl) adjustReadPartitions(numReadPartitions, numWritePartitions int32) int32 {
+	if numReadPartitions <= numWritePartitions {
+		return numWritePartitions
+	}
+	// check the backlog of the drained partitions
+	for i := numReadPartitions - 1; i >= numWritePartitions; i-- {
+		resp, err := a.matchingClient.DescribeTaskList(a.ctx, &types.MatchingDescribeTaskListRequest{
+			DomainUUID: a.taskListID.GetDomainID(),
+			DescRequest: &types.DescribeTaskListRequest{
+				TaskListType: a.taskListType,
+				TaskList: &types.TaskList{
+					Name: a.taskListID.GetPartition(int(i)),
+					Kind: types.TaskListKindNormal.Ptr(),
+				},
+				IncludeTaskListStatus: true,
+			},
+		})
+		if err != nil {
+			a.logger.Error("failed to get task list backlog", tag.Error(err))
+			break
+		}
+		nw := int32(1)
+		if resp.PartitionConfig != nil {
+			nw = resp.PartitionConfig.NumWritePartitions
+		}
+		// in order to drain a partition, 2 conditions need to be met:
+		// 1. the backlog size is 0
+		// 2. no task is being added to the partition, which is guaranteed to be true if the partition knows that the number of write partition is less or equal to its partition ID
+		if resp.TaskListStatus.GetBacklogCountHint() == 0 && nw <= i {
```

Comment:
since we don't take any action if `nw <= i` let's change the for loop to incorporate that to avoid unnecessary DescribeTaskList calls. 

```
nw := int32(1)
if resp.PartitionConfig != nil {
	nw = resp.PartitionConfig.NumWritePartitions
}
limit := max(numWritePartitions, nw)
for i := numReadPartitions; i > limit; i-- {
   partition := i -1
...
}
```